### PR TITLE
Add generics aware `isAssignableFrom` to `Argument`

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

File: core/src/main/java/io/micronaut/core/type/Argument.java [hunk 1]
```diff
@@ -174,6 +174,37 @@ default boolean isAssignableFrom(@NonNull Class<?> candidateType) {
         return getType().isAssignableFrom(Objects.requireNonNull(candidateType, "Candidate type cannot be null"));
     }
 
+    /**
+     * Checks if the argument can be assigned to this argument.
+     * @param candidateArgument The candidate argument
+     * @return True if it is assignable from.
+     * @since 3.0.0
+     */
+    default boolean isAssignableFrom(@NonNull Argument candidateArgument) {
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/type/Argument.java]: Should be `Argument<?>` with the wild card

File: core/src/main/java/io/micronaut/core/type/Argument.java [hunk 2]
```diff
@@ -174,6 +174,37 @@ default boolean isAssignableFrom(@NonNull Class<?> candidateType) {
...
+            return candidateArgumentTypeParameters.length >= 0;
+        }
+        if (candidateArgumentTypeParameters.length == 0) {
+            // Wildcard
+            return false;
+        }
+        for (int i = 0; i < typeParameters.length; i++) {
+            Argument typeParameter = typeParameters[i];
+            Argument candidateArgumentTypeParameter = candidateArgumentTypeParameters[i];
+            if (!typeParameter.isAssignableFrom(candidateArgumentTypeParameter)) {
```
Comment (graemerocher) [inline core/src/main/java/io/micronaut/core/type/Argument.java:201]: Do we need to check `isTypeVariable`?

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/micronaut/core/type/Argument.java (hunk 2 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/type/Argument.java:201]: We know it's coming from the type parameters so I don't think so

File: core/src/main/java/io/micronaut/core/type/Argument.java (hunk 1 above)
Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/type/Argument.java]: Added

**Pull request merged**